### ISSUE TITLE
Support HTTP/2.0

### ISSUE THREAD

**Issue by clue** [reactphp/http#99]

**Comment by maciejmrozinski** [comment]
Hi,

Do You have any WIP for this, plans, ideas how to implement this? Or maybe a good point of start will be after 1.0 release when internal API and classes will be stabilized? Any info appreciated :)

**Comment by clue** [comment]
The main focus is currently on getting the v1.0.0 out. As such, there are currently no immediate plans to build this from our end (no demand at the moment), but I suppose we would be really happy to accept PRs :+1:

**Comment by WyriHaximus** [comment]
@clue I would like to see HTTP/2 support after 1.0.0 is out. But we need certain features for that first.

**Comment by maciejmrozinski** [comment]
Ok, so i will get back to topic after 1.0 :)

**Comment by cordoval** [comment]
how does this have to do with version 2? i thought version 2 is an nginx thing rather?

**Comment by clue** [comment]
@cordoval Yes, there are a number of use cases where you can use this project without requiring HTTP/2.0. For example, if you're using this project behind a reverse proxy (such as nginx) that is capable of speaking HTTP/2.0 to the client and forwards this as HTTP/1.x to this library, then you can probably live without native HTTP/2.0 support here.

However, we still want to look into implementing HTTP/2.0 here as well so that this can also be used without a reverse proxy 👍  In the meantime, there are plenty of HTTP/2.0 proxies that can be used as above 👍

**Comment by WyriHaximus** [comment]
@clue && @cordoval in fact most of my own sites are running `react/http` behind `nginx` that handles HTTP/2.0 and certificates for all the different sites hosted on that box.

**Comment by cordoval** [comment]
nice man, good to know

hey have you news where is igorw? 👴 just curious

**Comment by valzargaming** [comment]
Are there any plans to support HTTP/3?

**Comment by clue** [comment]
@valzargaming I can definitely see us implementing HTTP/3 in the future, but baby steps first! In the meantime, using a reverse proxy to handle HTTP/2 and HTTP/3 is perfectly reasonable for many uses, but we definitely welcome contributions to help us land full support here as well :+1:

**Comment by SimonFrings** [comment]
It seems like this ticket is open for quite a while now. I want to avoid issues and pull request laying around for too long, but this depends on each topic. I think HTTP/2 support will come in the future, but I don't think there's something scheduled yet.

With that said we could close this ticket for now and reopen it once there are more detailed plans for HTTP/2 support, or we could leave it open, I will leave this decision to @clue as he opened this up :+1:

**Comment by WyriHaximus** [comment]
> It seems like this ticket is open for quite a while now. I want to avoid issues and pull request laying around for too long, but this depends on each topic. I think HTTP/2 support will come in the future, but I don't think there's something scheduled yet.

IMHO we still want to do this but need to do some internal refactoring first. It is very much on my list of things to add.

**Comment by masakielastic** [comment]
While experimenting with HTTP/2 on top of ReactPHP's socket component, I recently put together a very small educational example that demonstrates frame-level HTTP/2 communication implemented directly in PHP:

https://github.com/orgs/reactphp/discussions/616

This intentionally avoids external HTTP/2 libraries and only implements the minimal pieces required for a simple request. It is obviously incomplete and not intended for production use, but it helped me explore what a minimal HTTP/2 client might look like when built directly on ReactPHP's socket layer.

This made me curious about the long-term direction for HTTP/2 support in ReactPHP.

From an ecosystem perspective, there seem to be a few possible approaches:

1. A minimal HTTP/2 implementation in pure PHP built on top of `react/socket`
2. A PHP implementation with optional native acceleration (e.g. bindings to libraries like nghttp2)
3. A separate experimental package outside of the main `react/http` component

I'm currently experimenting with both a minimal PHP implementation and a prototype extension based on nghttp2, but I'm not sure which direction would best align with ReactPHP's design philosophy and maintenance expectations.

I'd be very interested to hear what maintainers think about the preferred approach for HTTP/2 support in the ReactPHP ecosystem.
